Molstar Progressive Web Apps

agdturner opened the pull request:
# Description
Mol* things installable as Progressive Web Apps (PWAs).

This follows a discussion:
https://github.com/molstar/molstar/discussions/1390

PWAs are explained here:
https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps
https://en.wikipedia.org/wiki/Progressive_web_app

There are distinct PWAs for viewer, docking-viewer, mesoscale-explorer and some examples. Two scripts (pwa.js and pwa-clean.js) automate the code generation (pwa.js) and clean up (pwa-clean.js). For each PWA, an index-pwa.html is generated from the original index.html. The respective index.ts and index.tsx files were updated to add reference to these additional html files. The service worker and manifest files are (re)generated as part of the build.

Examples of the PWAs are linked below:
https://agdturner.github.io/molstar/build/viewer/index-pwa.html
https://agdturner.github.io/molstar/build/docking-viewer/index-pwa.html
https://agdturner.github.io/molstar/build/mesoscale-explorer/index-pwa.html
https://agdturner.github.io/molstar/build/examples/alpha-orbitals/index-pwa.html
https://agdturner.github.io/molstar/build/examples/basic-wrapper/index-pwa.html
https://agdturner.github.io/molstar/build/examples/lighting/index-pwa.html
https://agdturner.github.io/molstar/build/examples/proteopedia-wrapper/index-pwa.html

The above PWAs are served from GitHub Pages. Once installed by a user as a PWA, they can work offline to use some Mol* functionality.

## Actions

- [y] Added description of changes to the `[Unreleased]` section of `CHANGELOG.md`
- [n] Updated headers of modified files (I don't think there was anything that should have been modified...)
- [y] Added my name to `package.json`'s `contributors`
- [n] (Optional but encouraged) Improved documentation in `docs` (Maybe there is something wanted here, but this is essentially only a sort of wrapper around the existing offering...)

arose reviewed (changes requested):
Hi @agdturner, thanks for doing this!

- I think it is enough to do it for the Viewer app
- None of the files created by pwa.js should be committed
- The path used in the manifest and elsewhere should be relative
- The pwa.js script should only write to the build/ directory
- How about adding support for file handlers?

agdturner commented:
> Hi @agdturner, thanks for doing this!
> 
> * I think it is enough to do it for the Viewer app
> * None of the files created by pwa.js should be committed
> * The path used in the manifest and elsewhere should be relative
> * The pwa.js script should only write to the build/ directory
> * How about adding support for file handlers?

I've simplified things so the PWA is created only for the Viewer App. pwa.js compiles things directly into build. The paths are all relative. Good suggestion about the file handlers. I will add that now. Sorry I jumped the gun and re-requested a review before I had that bit done.

arose commented:
> I simplified so a PWA is only created only for the Viewer App. pwa.js now compiles things directly into build. Paths all changed to be relative. File handlers added.

thanks, can you please remove all the extra files from the repository

agdturner commented:
> > I simplified so a PWA is only created only for the Viewer App. pwa.js now compiles things directly into build. Paths all changed to be relative. File handlers added.
> 
> thanks, can you please remove all the extra files from the repository, we don't commit any builds

I created a new branch called "build" and removed the build directory from the "master" branch...

agdturner commented:
> > I simplified so a PWA is only created only for the Viewer App. pwa.js now compiles things directly into build. Paths all changed to be relative. File handlers added.
> 
> thanks, can you please remove all the extra files from the repository, we don't commit any builds



> > I simplified so a PWA is only created only for the Viewer App. pwa.js now compiles things directly into build. Paths all changed to be relative. File handlers added.
> 
> thanks, can you please remove all the extra files from the repository, we don't commit any builds

I removed the build directory and created a new branch for deployment.

Here are instructions for getting Mol* served out via GitHub Pages and installable as a PWA:
1. Fork and clone molstar.
2. Create a new branch.
3. Checkout the new branch.
4. Build.
5. Edit .gitignore so that the build directory is pushed to the branch.
6. Push.
7. On GitHub in settings serve out the branch via GitHub Pages
8. Navigate to the following replacing <user> with the users GitHub username:
https://<user>.github.io/molstar/build/viewer/index-pwa.html
9. Install the PWA (how to do this depends on your browser and underlying platform).

I find installation as a PWA easiest with Chrome.

Perhaps a good idea to test those instructions and add them maybe to the readme?

agdturner commented:
Hi @arose, To make things cleaner better, I reworked this and created a replacement PR: https://github.com/molstar/molstar/pull/1436

HTH

Andy

agdturner commented:
In an attempt to make things easier, I deleted the pwa branch and replaced it. However this closed the open Pull Request. Sorry. I created a new PR after testing the PWA installs from some transpiled TypeScript:
https://github.com/molstar/molstar/pull/1454

Thanks again @arose for reviewing.

dsehnal commented:
We can close this now, right?

arose commented:
yes
